.agents/docs/fray-migration.md

```diff
@@ -0,0 +1,237 @@
+
+```python
+ctx = fray_job_ctx()
+future = ctx.run(remote_function, arg1, arg2)
+assert ctx.get(future) == 10
+```
+
+The simplest cases involve bare ray.remote calls and can be handled as above.
+
+More complex cases require analysis.
+
+#### Runtime Environment
+
+If a runtime_environment is specified in order to install packages or set environment variables,
+launch a new Job with the appropriate JobRequest:
+
+```python
+ray.remote(runtime_env=build_runtime_env_for_packages(...))
+def foo():
+
+```
+
+Becomes:
+
+```python
+from fray import JobRequest
+
+# foo just uses packages as needed
+def foo():
+    import package_x
+    import package_y
+
+def foo_caller():
+    request = JobRequest(
+        name="foo",
+        entrypoint=Entrypoint(callable=foo),
+        resources=ResourceConfig(replicas=1),
+        environment=create_environment(pip_packages=["package_x", "package_y"]),
+    )
+    ctx = current_cluster()
+    job_id = ctx.launch(request)
+    ctx.wait(job_id)
+
+### Resource requirements
+
+Fray breaks up job and task scheduling into separate concerns. We have 2 places this impacts our changes:
+
+#### Head node scheduling
+
+Scheduling on "head" node. This is typically used to place a controller process
+which should not be preempted. In Fray, this is expressed by putting a
+requirement for: "non-premptible" in the ResourceConfig for a JobRequest.
+
+#### TPU scheduling
+
+TPUs should be scheduled like any other job, using a JobRequest. The TPU type
+and slice size are specified in the TpuConfig section of the ResourceConfig. For
+"multi-slice" operation, specify replicas > 1:
+
+```
+
+ resource_config = ResourceConfig(
+        cpu=1,
+        ram="16",
+        disk="10g",
+        device=TpuConfig(type="v5litepod-4"),
+        replicas=1,
+        regions=["eu-west4"],
+    )
+
+    job_request = JobRequest(
+        name="vllm-inference-pool",
+        entrypoint=Entrypoint(
+            callable=vllm_server_worker,
+            function_args={
+                "model": self.config.model_config,
+                "request_queue": self.request_queue,
+                "response_queue": self.response_queue,
+            },
+        ),
+        resources=resource_config,
+        environment=create_environment(),
+    )
+
+    ctx = current_cluster()
+    job_id = ctx.launch(request)
+    ctx.wait(job_id)
+```
+
+### Ray actors
+
+TBD
+
+### Inference tasks
+
+TBD
+
+## Files Requiring Migration
+
+This section catalogs all files in `lib/marin/` that use Ray, organized by migration strategy.
+
+### Convert to Zephyr
+
+#### `lib/marin/src/marin/generation/inference.py`
+Text generation inference using Ray Data pipelines for vLLM inference on TPUs.
+**Approach:** Convert Ray Data pipeline to Zephyr for concurrent data processing.
+
+#### `lib/marin/src/marin/datashop/pipeline.py`
+MEDU data generation pipeline using Ray Data for concurrent vLLM inference with placement group scheduling.
```

Comment:
probably ok to delete medu at this point

Reply:
good to know, that will save me some pain